libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -192,24 +189,50 @@ proc broadcast*(
       else:
         libp2p_pubsub_broadcast_prune.inc(npeers, labelValues = ["generic"])
 
+proc broadcast*(
+  p: PubSub,
+  sendPeers: auto, # Iteratble[PubSubPeer]
+  msg: RPCMsg) {.raises: [Defect].} =
+  ## Attempt to send `msg` to the given peers
+
+  p.updateBroadcastMetrics(msg, sendPeers.len.int64)
   trace "broadcasting messages to peers",
     peers = sendPeers.len, msg = shortLog(msg)
 
   if anyIt(sendPeers, it.hasObservers):
     for peer in sendPeers:
-      p.send(peer, msg)
+      asyncSpawn p.send(peer, msg)
   else:
     # Fast path that only encodes message once
     let encoded = encodeRpcMsg(msg, p.anonymize)
     for peer in sendPeers:
       asyncSpawn peer.sendEncoded(encoded)
 
+proc cancellableBroadcast*(
+  p: PubSub,
+  sendPeers: auto, # Iteratble[PubSubPeer]
+  msg: RPCMsg): Table[PeerId, Future[void]] {.raises: [Defect].} =
+  ## Attempt to send `msg` to the given peers
+
+  p.updateBroadcastMetrics(msg, sendPeers.len)
+  trace "broadcasting messages to peers",
+    peers = sendPeers.len, msg = shortLog(msg)
+
+  if anyIt(sendPeers, it.hasObservers):
+    for peer in sendPeers:
+      result[peer.peerId] = p.send(peer, msg)
+  else:
+    # Fast path that only encodes message once
+    let encoded = encodeRpcMsg(msg, p.anonymize)
+    for peer in sendPeers:
+      result[peer.peerId] = peer.sendEncoded(encoded)
```

Comment:
doesn't this already synchronously put bytes in the send buffer?

Reply:
With mplex, that will indeed go synchronously through every layer, down to chronos
I thought that chronos supported write cancellation, apparently not, quoting cheatfate:
> There is a big problem with write and i wish not to solve it today or tomorrow, because write is made using queue, so cancelling write in the middle should remove this write call from the queue, but it will break cross-platform compatibility, because on Windows this queue is in system, so i can't freely remove write call from the middle of the queue...

So we could add cancellation support to linux at least

With yamux, there is an intermediary queue before chronos (since there is per-stream backpressure), that does support cancellation:
https://github.com/status-im/nim-libp2p/blob/351bda2b56e4389a91fb42199621b3617bdbbdd3/libp2p/muxers/yamux/yamux.nim#L285

So this PR only makes sense if we get chronos support for write cancellation on linux, or enable yamux in nimbus :/

Reply:
Getting TCP cancellation support in chronos seems less complex than expected: https://github.com/status-im/nim-chronos/pull/353
I'll wait for reviews to be sure